### PR TITLE
[SPARK-45427][CORE] Add RPC SSL settings to SSLOptions and SparkTransportConf

### PR DESCRIPTION
**Pull request by hasnain-db**
### What changes were proposed in this pull request?

This PR adds the options added in https://github.com/apache/spark/pull/43220  to `SSLOptions` and `SparkTransportConf`.

By adding it to the `SSLOptions` we can support inheritance of options, so settings for the UI and RPC SSL settings can be shared as much as possible. The `SparkTransportConf` changes are needed to support propagating these settings.

I also make some changes to `SecurityManager` to log when this feature is enabled, and make the existing `spark.network.crypto` options mutually exclusive with this new settings (it would just involve double encryption then).

Lastly, make these flags propagate down to when executors are launched, and allow the passwords to be sent via environment variables (similar to how it's done for an existing secret). This ensures they are not visible in plaintext, but also ensures they are available at executor startup (otherwise it can never talk to the driver/worker)

### Why are the changes needed?

The propagation of these options are needed for the RPC functionality to work

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

CI

Added some unit tests which I verified passed:

```
build/sbt
> project core
> testOnly org.apache.spark.SparkConfSuite org.apache.spark.SSLOptionsSuite org.apache.spark.SecurityManagerSuite org.apache.spark.deploy.worker.CommandUtilsSuite
``` 

The rest of the changes and integration were tested as part of https://github.com/apache/spark/pull/42685

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by hasnain-db** [comment]
cc: @mridulm @JoshRosen

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 1]
```diff
@@ -258,18 +265,46 @@ private[spark] class SecurityManager(
     isUserInACL(user, modifyAcls, modifyAclsGroups)
   }
 
+  /**
+   * Check to see if authentication is enabled and SSL for RPC is disabled (as they are mutually
+   * exclusive)
+   * @return Whether authentication is enabled and SSL for RPC is disabled
+   */
+  private def isAuthenticationEnabledAndSslRpcDisabled(): Boolean = {
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala:273]: Why is authentication mutually exclusive from rpc ssl ?
For example, external shuffle service uses it to validate identity of the incoming request.

IMO let us keep SSL for securing over-the-wire comm

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 2]
```diff
@@ -258,18 +265,46 @@ private[spark] class SecurityManager(
...
    * @return true if authentication is enabled, otherwise false
    */
-  def isAuthenticationEnabled(): Boolean = authOn
+  def isAuthenticationEnabled(): Boolean = isAuthenticationEnabledAndSslRpcDisabled
 
   /**
    * Checks whether network encryption should be enabled.
    * @return Whether to enable encryption when connecting to services that support it.
    */
   def isEncryptionEnabled(): Boolean = {
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala:297]: Fix the warning in `SparkEnv` as well ?

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 3]
```diff
@@ -739,6 +739,9 @@ private[spark] object SparkConf extends Logging {
     (name.startsWith("spark.auth") && name != SecurityManager.SPARK_AUTH_SECRET_CONF) ||
     name.startsWith("spark.rpc") ||
     name.startsWith("spark.network") ||
+    // We need SSL configs to propagate as they may be needed for RPCs.
+    // Passwords are propagated separately though.
+    (name.startsWith("spark.ssl") && !name.contains("Password")) ||
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/SparkConf.scala:744]: Dont you need the password in order to load the keys ?

**Review by mridulm — COMMENTED** [review]
w.r.t the config namespace, we currently have:
* `spark.ssl` for UI related SSL configs.
* rpc IO is currently under `spark.network.crypto`
* `spark.rpc.` for various rpc related configs, but they dont deal with security.
* `spark.io.encryption` which deals with block encryption.

IMO we should namespace to `spark.network` and add `ssl` there, instead of moving these configs for rpc to `spark.ssl`.

Thoughts @JoshRosen, @hasnain-db ?

File: core/src/main/scala/org/apache/spark/SSLOptions.scala [hunk 4]
```diff
@@ -134,12 +157,44 @@ private[spark] case class SSLOptions(
     supported
   }
 
+  /**
+   *
+   * @return
+   */
```
Comment (beliefer) [inline core/src/main/scala/org/apache/spark/SSLOptions.scala:163]: Please supply the comment or remove it.

File: core/src/main/scala/org/apache/spark/SSLOptions.scala [hunk 5]
```diff
@@ -152,15 +207,25 @@ private[spark] object SSLOptions extends Logging {
...
+   * $ - `[ns].trustStoreReloadingEnabled` - enables or disables using a trust-store
+   * that reloads its configuration when the trust-store file on disk changes
+   * $ - `[ns].trustStoreReloadIntervalMs` - the interval, in milliseconds, the
+   * trust-store will reload its configuration
+   * $ - `[ns].openSslEnabled` - enables or disables using an OpenSSL implementation
+   * (if available on host system), requires certChain and keyFile arguments
    * $ - `[ns].protocol` - a protocol name supported by a particular Java version
    * $ - `[ns].enabledAlgorithms` - a comma separated list of ciphers
+   * $ - `[ns].dangerouslyFallbackIfKeysNotPresent` - whether to fallback to unencrypted
+   *                                                  communication if keys are not present.
```
[the remark this answers is not in the document]
Comment (hasnain-db) [reply core/src/main/scala/org/apache/spark/SSLOptions.scala:228]: this is helpful in certain deployment settings where a key might not be available at node startup but is available before it needs to communicate to other machines. Happy to keep this out of OSS spark though, I don't feel too strongly about this.

File: core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala [hunk 6]
```diff
@@ -38,12 +56,14 @@ object SparkTransportConf {
    *                       This restriction will only occur if these properties are not already set.
    * @param role           optional role, could be driver, executor, worker and master. Default is
    *                      [[None]], means no role specific configurations.
+   * @param sslOptions SSL config options
    */
-  def fromSparkConf(
+  def fromSparkConfWithSslOptions(
```
[the remark this answers is not in the document]
Comment (hasnain-db) [reply core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala:61]: Tests fail saying that this breaks binary compatibility. I initially went with that but changed it for that reason

File: core/src/main/scala/org/apache/spark/SecurityManager.scala (hunk 1 above)
Comment (hasnain-db) [reply core/src/main/scala/org/apache/spark/SecurityManager.scala:273]: Thanks! I had misread the docs earlier and thought authentication required encryption, but re-reading now I realize it's the other way around. I can adjust. While TLS can handle authentication too (with the right trust stores) it makes sense to keep existing auth mechanisms working over TLS.

Just to confirm, we do want to keep the legacy encryption and TLS mutually exclusive right? I don't see a benefit to double encrypting (just overhead)

File: core/src/main/scala/org/apache/spark/SecurityManager.scala (hunk 2 above)
Comment (hasnain-db) [reply core/src/main/scala/org/apache/spark/SecurityManager.scala:297]: ah, good catch: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SparkEnv.scala#L310-L315 should not warn if TLS is enabled I imagine. I'll try to fix this.

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 3 above)
Comment (hasnain-db) [reply core/src/main/scala/org/apache/spark/SparkConf.scala:744]: We do, and we pass them over environment variables. `isExecutorStartupConf` is used for arguments that are passed in the command line invocation to start the executor. If we passed the password there, anyone on the system using `ps` could see the password in plaintext, which we don't want. There is an existing test which fails, due to exactly this bug, that I had to adjust in this PR.

**Comment by hasnain-db** [comment]
> w.r.t the config namespace, we currently have ...
> IMO we should namespace to spark.network and add ssl there, instead of moving these configs for rpc to spark.ssl.

My main preference here was to use the existing `spark.ssl` settings because there is a nice interop with how SSL works for the UI: we can reuse the same keystore and trustore and protocol settings and have things be consistent. I agree it's a little confusing as it's not under the network settings, but I don't see it being a stretch for a user to expect SSL settings to be under `spark.ssl` - that's naively where I checked initially when looking for RPC SSL support.

What do you think? I'm happy to defer to you and @JoshRosen for the final call here.

File: core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala:61]: We can add it to mima exclusion rules as the change is preserving source compatibility.

File: core/src/main/scala/org/apache/spark/SecurityManager.scala (hunk 1 above)
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/SecurityManager.scala:273]: Agree, does not make sense to do existing encryption over TLS :-)

File: core/src/main/scala/org/apache/spark/SparkConf.scala (hunk 3 above)
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/SparkConf.scala:744]: Sounds good, thanks for clarifying

File: core/src/main/scala/org/apache/spark/SSLOptions.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (hasnain-db) [reply core/src/main/scala/org/apache/spark/SSLOptions.scala:228]: removed.

**Comment by hasnain-db** [comment]
addressed part of the feedback. Still waiting for @JoshRosen to also chime in with his thoughts re: which config namespace to use.

**Review by JoshRosen — COMMENTED** [review]
I don't have super strong feelings about the configuration namespacing: I'm okay with any option as long as our docs help users to chart a clear course for common deployment scenarios.

Some assorted thoughts:

- As we think through this, it might be useful to also review some of the proposed documentation updates from the main PR at https://github.com/apache/spark/pull/42685/files#diff-2c506cc756a5333f8fb5acb850d443980706602048f9ddb3dc515abacbfbce6a (in `security.md`) to see if we think these scenarios make sense.
  - Re-reading this now, I think we might want to re-work the page's introduction to offer more clarity on the different options for encryption so users have a clearer understanding of which options are mutually-exclusive. For example, we should clarify whether the existing shared secret authentication can be used with SSL.
  - We might want to re-work the intro to more clearly state that there are two mutually-exclusive choices for RPC encryption, and might want to leave some breadcrums cross-linking some sub-sections to reiterate this point.
- There may be some advantages to the hierarchical namespacing of SSL configurations for the sake of giving users a single place to configure `enabledAlgorithms` (e.g. if a user has a compliance protocol that allows only certain algorithms then they shouldn't have to repeat themselves and keep multiple configurations in sync (although I suppose we could have a separate configuration that just inherits its default from the main SSL options)).
- In the long run, do we expect that new users would choose to use non-SSL RPC encryption mechanisms? Or would they configure SSL everywhere? Do we expect the existing non-UI encryption configurations to eventually be deprecated (SASL already is)? I'm wondering if we use some long-term vision  of a future end state to guide us here.

**Comment by hasnain-db** [comment]
@JoshRosen I updated the docs PR to hopefully add more clarity around the various encryption options (and, to your request, clarified how they overlap with the authentication setting).

> There may be some advantages to the hierarchical namespacing of SSL configurations...

Agreed.  If we reuse the `spark.ssl.rpc` namespace as is done in the PR the implementation is also much easier -- but we can also make the settings inheritance work if we use a different namespace for it - implementation complexity shouldn't hinder UX IMO.

>  I'm wondering if we use some long-term vision of a future end state to guide us here.

I don't have enough context to make a call here and defer to @mridulm . I will say that as a random observer here who hasn't had to manage too many spark deployments just yet, I feel like the current encryption mechanism has a strong point in its favor: it's quite simple to configure just a shared secret. SSL requires provisioning keys and certs which is always more work.

**Comment by mridulm** [comment]
I was trying to see if `spark.network.ssl` made more sense, and why we have `spark.ssl` and not `spark.ui.ssl` (given current use is for UI).

We used to have more ssl related configs in the past under that namespace - `spark.ssl.akka`, `spark.ssl.fs` (file server), etc ... with the intent to keep `spark.ssl` as the namespace for various module's ssl config.

Given this, even though today it is just UI, it does make more sense to go back to that approach.
Let us keep it as proposed in this PR.

File: core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala [hunk 7]
```diff
@@ -31,19 +31,21 @@ object SparkTransportConf {
...
-   * @param module the module name
+   * @param _conf          the [[SparkConf]]
+   * @param module         the module name
    * @param numUsableCores if nonzero, this will restrict the server and client threads to only
    *                       use the given number of cores, rather than all of the machine's cores.
    *                       This restriction will only occur if these properties are not already set.
    * @param role           optional role, could be driver, executor, worker and master. Default is
-   *                      [[None]], means no role specific configurations.
+   *                       [[None]], means no role specific configurations.
+   * @param sslOptions SSL config options
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/network/netty/SparkTransportConf.scala:41]: nit: Remove white space changes ?

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 8]
```diff
@@ -269,9 +276,24 @@ private[spark] class SecurityManager(
...
+    if (sparkConf.get(Network.NETWORK_CRYPTO_ENABLED) || sparkConf.get(SASL_ENCRYPTION_ENABLED)) {
+      if (rpcSSLOptions.enabled) {
+        logWarning("Network encryption disabled as RPC SSL encryption is enabled")
+        false
+      } else {
+        true
+      }
+    } else {
+      false
+    }
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala:288]: nit:

```suggestion
    val encryptionEnabled = sparkConf.get(Network.NETWORK_CRYPTO_ENABLED) || sparkConf.get(SASL_ENCRYPTION_ENABLED)
    if (encryptionEnabled && rpcSSLOptions.enabled) {
      logWarning("Network encryption disabled as RPC SSL encryption is enabled")
      false
    } else 
      encryptionEnabled
    }
```

**Review by mridulm — APPROVED** [review]
Just a couple of nits, looks good to me

**Comment by hasnain-db** [comment]
test failure looks unrelated, retrying for the third time

**Comment by hasnain-db** [comment]
`org.apache.spark.sql.kafka010.KafkaSourceStressSuite` is failing on this and the other outstanding PR, and seems to be unrelated to the changes here. I wonder if that is related to the new kafka version bump since it looks like it also failed on https://github.com/apache/spark/actions/runs/6497942586/job/17648192318

**Comment by mridulm** [comment]
The failures are not related.
Merging to master.
Thanks for working on this @hasnain-db !
Thanks for the review @JoshRosen :-)

**Pull request closed without merging**